<a href="https://colab.research.google.com/github/leeryleprechaun/MSDS_600_W5/blob/main/Week_5_assignment_starter.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# DS Automation Assignment

Using our prepared churn data from week 2:
- use pycaret to find an ML algorithm that performs best on the data
    - Choose a metric you think is best to use for finding the best model; by default, it is accuracy but it could be AUC, precision, recall, etc. The week 3 FTE has some information on these different metrics.
- save the model to disk
- create a Python script/file/module with a function that takes a pandas dataframe as an input and returns the probability of churn for each row in the dataframe
    - your Python file/function should print out the predictions for new data (new_churn_data.csv)
    - the true values for the new data are [1, 0, 0, 1, 0] if you're interested
- test your Python module and function with the new data, new_churn_data.csv
- write a short summary of the process and results at the end of this notebook
- upload this Jupyter Notebook and Python file to a Github repository, and turn in a link to the repository in the week 5 assignment dropbox

*Optional* challenges:
- return the probability of churn for each new prediction, and the percentile where that prediction is in the distribution of probability predictions from the training dataset (e.g. a high probability of churn like 0.78 might be at the 90th percentile)
- use other autoML packages, such as TPOT, H2O, MLBox, etc, and compare performance and features with pycaret
- create a class in your Python module to hold the functions that you created
- accept user input to specify a file using a tool such as Python's `input()` function, the `click` package for command-line arguments, or a GUI
- Use the unmodified churn data (new_unmodified_churn_data.csv) in your Python script. This will require adding the same preprocessing steps from week 2 since this data is like the original unmodified dataset from week 1.

In [ ]:
!python3 --version

Python 3.7.13


Ok, so after figthing with to pycaret installer for 2+ hours I found and article that suggests that it does not work on python past version 3.7, I have 3.9 installed on my computer. I am going to try using google colab and see if it works now.

In [1]:
!pip install pycaret

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 320 kB 14.3 MB/s 
     |████████████████████████████████| 17.8 MB 405 kB/s 
     |████████████████████████████████| 120 kB 48.5 MB/s 
     |████████████████████████████████| 1.3 MB 60.3 MB/s 
     |████████████████████████████████| 6.8 MB 64.0 MB/s 
     |████████████████████████████████| 262 kB 68.0 MB/s 
     |████████████████████████████████| 56 kB 2.5 MB/s 
     |████████████████████████████████| 167 kB 75.7 MB/s 
     |████████████████████████████████| 88 kB 7.3 MB/s 
     |████████████████████████████████| 2.0 MB 61.0 MB/s 
     |████████████████████████████████| 1.7 MB 48.1 MB/s 
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Installing backend dependencies ... done
    Preparing wheel metadata ... done
     |████████████████████████████████| 1.3 MB 65.5 MB/s 
     |████████████████████████████████| 

In [1]:
!pip install tf-estimator-nightly

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 438 kB 14.3 MB/s 
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
tensorflow 2.8.0+zzzcolab20220506162203 requires numpy>=1.20, but you have numpy 1.19.5 which is incompatible.
tensorflow 2.8.0+zzzcolab20220506162203 requires tf-estimator-nightly==2.8.0.dev2021122109, but you have tf-estimator-nightly 2.10.0.dev2022052908 which is incompatible.


In [2]:
!pip install numpy

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [3]:
!pip install folium

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [4]:
!pip install imgaug

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [35]:
import pandas as pd

from google.colab import drive 
drive.mount('/content/gdrive')

df = pd.read_csv('/content/gdrive/MyDrive/prepped_churn_data.csv', index_col='customerID')
df

Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).


tenure  PhoneService  Contract  PaymentMethod  MonthlyCharges  \
customerID                                                                  
7590-VHVEG       1             0         0              0           29.85   
5575-GNVDE      34             1         1              1           56.95   
3668-QPYBK       2             1         0              1           53.85   
7795-CFOCW      45             0         1              2           42.30   
9237-HQITU       2             1         0              0           70.70   
...            ...           ...       ...            ...             ...   
6840-RESVB      24             1         1              1           84.80   
2234-XADUH      72             1         1              3          103.20   
4801-JZAZL      11             0         0              0           29.60   
8361-LTMKD       4             1         0              1           74.40   
3186-AJIEK      66             1         2              2          105.65   

            TotalCharges  Churn  
customerID                       
7590-VHVEG        1.4925      0  
5575-GNVDE       94.4750      0  
3668-QPYBK        5.4075      1  
7795-CFOCW       92.0375      0  
9237-HQITU        7.5825      1  
...                  ...    ...  
6840-RESVB       99.5250      0  
2234-XADUH      368.1450      0  
4801-JZAZL       17.3225      0  
8361-LTMKD       15.3300      1  
3186-AJIEK      342.2250      0  

[7032 rows x 7 columns]

In [6]:
pip install -U Jinja2

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 133 kB 14.7 MB/s 
  Attempting uninstall: Jinja2
    Found existing installation: Jinja2 2.11.3
    Uninstalling Jinja2-2.11.3:
      Successfully uninstalled Jinja2-2.11.3
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
google-colab 1.0.0 requires requests~=2.23.0, but you have requests 2.27.1 which is incompatible.
flask 1.1.4 requires Jinja2<3.0,>=2.10.1, but you have jinja2 3.1.2 which is incompatible.
datascience 0.10.6 requires folium==0.2.1, but you have folium 0.8.3 which is incompatible.


In [36]:
from pycaret.classification import setup, compare_models, predict_model, save_model, load_model

Wow! This was a struggle, I spent several hours getting this to work, was never able to get it to work in jupyter lab becuase my version of python does not support pycaret. I was finally able to get it to run on google collab which has an earlier version of python. I then had a number of issues getting all the other files to work on google colab. Finally figured it out though!

In [51]:
automl = setup(df, target='Churn', preprocess=False)

Description             Value
0              session_id              7519
1                  Target             Churn
2             Target Type            Binary
3           Label Encoded              None
4           Original Data         (7032, 7)
5          Missing Values             False
6        Numeric Features                 3
7    Categorical Features                 3
8   Transformed Train Set         (4922, 6)
9    Transformed Test Set         (2110, 6)
10     Shuffle Train-Test              True
11    Stratify Train-Test             False
12         Fold Generator   StratifiedKFold
13            Fold Number                10
14               CPU Jobs                -1
15                Use GPU             False
16         Log Experiment             False
17        Experiment Name  clf-default-name
18                    USI              0206
19          Fix Imbalance             False
20   Fix Imbalance Method             SMOTE

In [52]:
automl[3]

tenure PhoneService Contract PaymentMethod  MonthlyCharges  \
customerID                                                               
7590-VHVEG     1.0            0        0             0       29.850000   
5575-GNVDE    34.0            1        1             1       56.950001   
3668-QPYBK     2.0            1        0             1       53.849998   
7795-CFOCW    45.0            0        1             2       42.299999   
9237-HQITU     2.0            1        0             0       70.699997   
...            ...          ...      ...           ...             ...   
6840-RESVB    24.0            1        1             1       84.800003   
2234-XADUH    72.0            1        1             3      103.199997   
4801-JZAZL    11.0            0        0             0       29.600000   
8361-LTMKD     4.0            1        0             1       74.400002   
3186-AJIEK    66.0            1        2             2      105.650002   

            TotalCharges  
customerID                
7590-VHVEG      1.492500  
5575-GNVDE     94.474998  
3668-QPYBK      5.407500  
7795-CFOCW     92.037498  
9237-HQITU      7.582500  
...                  ...  
6840-RESVB     99.525002  
2234-XADUH    368.144989  
4801-JZAZL     17.322500  
8361-LTMKD     15.330000  
3186-AJIEK    342.225006  

[7032 rows x 6 columns]

In [53]:
best_model = compare_models()

Model  Accuracy     AUC  Recall   Prec.  \
lr                 Logistic Regression    0.7942  0.8335  0.5050  0.6412   
lda       Linear Discriminant Analysis    0.7932  0.8219  0.5065  0.6387   
ridge                 Ridge Classifier    0.7907  0.0000  0.4442  0.6532   
ada               Ada Boost Classifier    0.7907  0.8362  0.4872  0.6356   
gbc       Gradient Boosting Classifier    0.7905  0.8349  0.4826  0.6381   
knn             K Neighbors Classifier    0.7757  0.7740  0.4803  0.5943   
rf            Random Forest Classifier    0.7753  0.8023  0.4834  0.5930   
et              Extra Trees Classifier    0.7617  0.7801  0.4796  0.5579   
qda    Quadratic Discriminant Analysis    0.7536  0.8219  0.7321  0.5249   
svm                SVM - Linear Kernel    0.7515  0.0000  0.5026  0.6189   
dummy                 Dummy Classifier    0.7361  0.5000  0.0000  0.0000   
dt            Decision Tree Classifier    0.7282  0.6544  0.4881  0.4865   
nb                         Naive Bayes    0.7202  0.8070  0.7567  0.4815   

           F1   Kappa     MCC  TT (Sec)  
lr     0.5638  0.4318  0.4378     0.080  
lda    0.5637  0.4307  0.4365     0.017  
ridge  0.5278  0.3998  0.4126     0.022  
ada    0.5504  0.4175  0.4243     0.228  
gbc    0.5475  0.4151  0.4230     0.316  
knn    0.5306  0.3855  0.3897     0.153  
rf     0.5316  0.3860  0.3901     0.782  
et     0.5145  0.3581  0.3606     0.563  
qda    0.6110  0.4380  0.4513     0.016  
svm    0.4890  0.3418  0.3737     0.054  
dummy  0.0000  0.0000  0.0000     0.013  
dt     0.4867  0.3020  0.3024     0.036  
nb     0.5883  0.3921  0.4153     0.027

According to this analysis the Logistic Regression model will provide the best overall fit.

In [54]:
best_model

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=1000,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=7519, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [55]:
df.iloc[-2:-1].shape

(1, 7)

This is graddbing the last row of data

In [56]:
predict_model(best_model, df.iloc[-2:-1])

Model  Accuracy  AUC  Recall  Prec.   F1  Kappa  MCC
0  Logistic Regression       1.0    0     1.0    1.0  1.0    NaN  0.0

tenure  PhoneService  Contract  PaymentMethod  MonthlyCharges  \
customerID                                                                  
8361-LTMKD       4             1         0              1            74.4   

            TotalCharges  Churn  Label   Score  
customerID                                      
8361-LTMKD         15.33      1      1  0.5725

In [57]:
save_model(best_model, 'LR')

Transformation Pipeline and Model Successfully Saved


(Pipeline(memory=None,
          steps=[('dtypes',
                  DataTypes_Auto_infer(categorical_features=[],
                                       display_types=True, features_todrop=[],
                                       id_columns=[],
                                       ml_usecase='classification',
                                       numerical_features=[], target='Churn',
                                       time_features=[])),
                 ['trained_model',
                  LogisticRegression(C=1.0, class_weight=None, dual=False,
                                     fit_intercept=True, intercept_scaling=1,
                                     l1_ratio=None, max_iter=1000,
                                     multi_class='auto', n_jobs=None,
                                     penalty='l2', random_state=7519,
                                     solver='lbfgs', tol=0.0001, verbose=0,
                                     warm_start=False)]],
          verbose=

In [58]:
import pickle

with open('LR.pk', 'wb') as f:
  pickle.dump(best_model, f)

In [59]:
with open('LR.pk', 'rb') as f:
    loaded_model = pickle.load(f)

In [60]:
new_data = df.iloc[-2:-1].copy()
new_data.drop('Churn', axis=1, inplace=True)
loaded_model.predict(new_data)

array([1])

In [61]:
loaded_model = load_model('LR')

Transformation Pipeline and Model Successfully Loaded


In [62]:
predict_model(loaded_model, new_data)

tenure  PhoneService  Contract  PaymentMethod  MonthlyCharges  \
customerID                                                                  
8361-LTMKD       4             1         0              1            74.4   

            TotalCharges  Label   Score  
customerID                               
8361-LTMKD         15.33      1  0.5725

In [64]:
import pandas as pd

from google.colab import drive 
drive.mount('/content/gdrive')

df = pd.read_csv('/content/gdrive/MyDrive/new_churn_data.csv', index_col='customerID')
df

Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).


tenure  PhoneService  Contract  PaymentMethod  MonthlyCharges  \
customerID                                                                  
9305-CKSKC      22             1         0              2           97.40   
1452-KNGVK       8             0         1              1           77.30   
6723-OKKJM      28             1         0              0           28.25   
7832-POPKP      62             1         0              2          101.70   
6348-TACGU      10             0         0              1           51.15   

            TotalCharges  charge_per_tenure  
customerID                                   
9305-CKSKC        811.70          36.895455  
1452-KNGVK       1701.95         212.743750  
6723-OKKJM        250.90           8.960714  
7832-POPKP       3106.56          50.105806  
6348-TACGU       3440.97         344.097000

In [73]:
import pandas as pd
from pycaret.classification import predict_model, load_model

def load_data(filepath):
    """
    Loads churn data into a DataFrame from a string filepath.
    """
    df = pd.read_csv(filepath, index_col='customerID')
    return df


def make_predictions(df):
    """
    Uses the pycaret best model to make predictions on data in the df dataframe.
    """
    model = load_model('LR')
    predictions = predict_model(model, data=df)
    predictions.rename({'Label': 'Churn_prediction'}, axis=1, inplace=True)
    predictions['Churn_prediction'].replace({1: 'Churn', 0: 'No Churn'}, inplace=True)

    return predictions['Churn_prediction']


if __name__ == "__main__":
    df = load_data('/content/gdrive/MyDrive/new_churn_data.csv')
    predictions = make_predictions(df)
    print('predictions:')
    print(predictions)

Transformation Pipeline and Model Successfully Loaded
predictions:
customerID
9305-CKSKC       Churn
1452-KNGVK       Churn
6723-OKKJM    No Churn
7832-POPKP       Churn
6348-TACGU       Churn
Name: Churn_prediction, dtype: object


# Saving File to Github


# Summary

Before the summary I would like to note what a challenge this weeks assignmet was! A lot of the content is out of date, and nothing worked as it was supposed to this week. I ended writting the entire notebook in google colab because of all the version errors that exist in the Python 3.9 library that is loaded with Anaconda. I tried to roll back the version, run a virtual enviornment, really everything else I could. Thankfully google colab worked in the end.

After comparing a number of machine learning models based on the current churn training data set it was determined that using a Linear Regression model provided the best predictive model for the data. The linear regression model was then trained on the data set and used to predict churn for a set of 5 customers. Based on the model customers 9305-CKSKC, 1452-KNGVK, 7832-POPKP, and 6348-TACGU are likely to leave the phone plan. I would suggest that these customers be offered a discount as an incentive to stay with the company. It would also help if to receive the incentive the customers had to change to an automatic payment form, which appears to help retain customers according to the data set. 